<launch>

  <machine name="pi" address="PI" default="true" />
  <machine name="controller" address="Joseph-ROS" default="never" env-loader="/root/RoboCar/ROS_Workspace/scripts/controller_environment.sh" />

  <group ns="robocar">
    <rosparam param="ticks_meter">100</rosparam> <!-- int, ~200mm circumference, 20 ticks/rev -->

    <node machine="controller" pkg="robocar" name="joyctl" type="joyctl.py">
      <remap from="twist" to="differential/twist" />
    </node>

    <node machine="controller" pkg="image_view" type="image_view" name="image_view" >
      <remap from="image" to="camera/image_raw" />
    </node>

    <!-- differential transform, converts wheel encoders into odometry and a transform -->
    <node pkg="differential_drive" type="diff_tf.py" name="differential_transform">
      <rosparam param="rate">8.0</rosparam>
      <rosparam param="base_width">0.14747</rosparam> <!-- float meters-->

      <remap from="odom" to="base/odometry" /> 
      <remap from="tf" to="base/transform" /> 

      <remap from="lwheel" to="left_wheel/encoder" />
      <remap from="rwheel" to="right_wheel/encoder" />
    </node>

    <!-- left wheel PID controller -->
    <node pkg="differential_drive" type="pid_velocity" name="left_wheel_pid">
      <remap from="motor_cmd" to="left_wheel/motor_pwm" />
      <remap from="wheel_vel" to="left_wheel/velocity" />
      
      <remap from="wheel" to="left_wheel/encoder" />
      <remap from="wheel_vtarget" to="left_wheel/target_velocity" />

      <rosparam param="Kp">90</rosparam>
      <rosparam param="Ki">90</rosparam>
      <rosparam param="Kd">22</rosparam>

      <rosparam param="out_min">-100</rosparam>
      <rosparam param="out_max">100</rosparam>
      <rosparam param="rate">10</rosparam>
      <rosparam param="rolling_pts">5</rosparam>
      <rosparam param="timeout_ticks">2</rosparam>
    </node>

    <!-- right wheel PID controller -->
    <node pkg="differential_drive" type="pid_velocity" name="right_wheel_pid">
      <remap from="motor_cmd" to="right_wheel/motor_pwm" />
      <remap from="wheel_vel" to="right_wheel/velocity" />
      
      <remap from="wheel" to="right_wheel/encoder" />
      <remap from="wheel_vtarget" to="right_wheel/target_velocity" />

      <rosparam param="Kp">90</rosparam>
      <rosparam param="Ki">90</rosparam>
      <rosparam param="Kd">22</rosparam>

      <rosparam param="out_min">-100</rosparam>
      <rosparam param="out_max">100</rosparam>
      <rosparam param="rate">10</rosparam>
      <rosparam param="rolling_pts">5</rosparam>
      <rosparam param="timeout_ticks">2</rosparam>
    </node>

    <node pkg="differential_drive" type="twist_to_motors.py" name="differential_twist_to_velocity">
      <rosparam param="base_width">0.14747</rosparam>
      <rosparam param="rate">30</rosparam>
      
      <remap from="lwheel_vtarget" to="left_wheel/target_velocity" />
      <remap from="rwheel_vtarget" to="right_wheel/target_velocity" />
      
      <remap from="twist" to="differential/twist" />"
    </node>

    <node pkg="usb_cam" name="camera" type="usb_cam_node" output="screen">
      <rosparam param="framerate">15</rosparam>
      <rosparam param="image_width">176</rosparam> 
      <rosparam param="image_height">144</rosparam> 
      <rosparam param="pixel_format">yuyv</rosparam>
      <rosparam param="focus">0</rosparam>
    </node>

    <node pkg="robocar" name="serial" type="serial_node" output="screen"></node>

  </group>

</launch>
